### PR TITLE
Implement statuses, complete most castable support

### PR DESCRIPTION
**Pull request by baughj**
## Description
This PR adds statuses to Hybrasyl (e.g. buffs/debuffs), along with all the handling you would expect. It also completely implements castables (with the notable exception of scripted castables, which will be handled when needed). 

Along the way a lot of code was refactored or removed. There is a new static class `NumberCruncher` for performing formula / simple calculations for healing and damage. It also handles statuses.

## Related PRs
n/a

## Checklist
- [x] Tested and working locally
- [x] Reviewed
- [ ] Tested and working on dev server
- [ ] Deployed to staging

## How to QA
The following things should now be functional, in addition to damage spells:

- [x] Casting heal spells (`ioc`)
- [x] Casting damage spells with multitargeting (`srad gar`, `srad lamh`) 
- [x] Casting a spell involving a status (`armachd`, `puinsein`)
- [x] Removing a status on a creature (`ao puinsein`)

## Impacted Areas in Hybrasyl
Statuses
Castables
Targeting
Formula calculations

### CODE REVIEW

**Comment by baughj** [comment]
There is still a lot of QA needed, which I will be doing over the next day or so, but the work is effectively complete now.

File: HybrasylXML/XSD/Castables.xsd [hunk 1]
```diff
@@ -92,8 +93,8 @@
         <xs:list itemType="hyb:Class" />
       </xs:simpleType>
     </xs:attribute>
-    <xs:attribute name="Id" type="xs:int"/>
-    <xs:attribute name="Speed" type="xs:int"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Castables.xsd]: Why the change on types?

File: HybrasylXML/XSD/Castables.xsd [hunk 2]
```diff
@@ -162,7 +163,7 @@
   <xs:complexType name="Intent">
     <xs:attribute name="UseType" type="actions:SpellUseType" default="NoTarget" use="optional"/>
     <xs:attribute name="Radius" type="xs:unsignedByte" default="0" />
-    <xs:attribute name="Direction" type="actions:IntentDirection" default="Front" />
+    <xs:attribute name="Direction" type="actions:IntentDirection" default="None" use="optional"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Castables.xsd]: How is optional use a thing for this?

File: HybrasylXML/XSD/Status.xsd [hunk 3]
```diff
@@ -22,10 +22,11 @@
...
       <xs:element name="Script" type="hyb:String8" />
       <xs:element name="Effects" type="status:Effects" minOccurs="0" maxOccurs="1"/>
+      <xs:element name="ProhibitedMessage" type="hyb:String8" minOccurs="0" maxOccurs="1"/>
     </xs:all>
-    <xs:attribute name="Duration" use="optional" />
-    <xs:attribute name="Tick" use="optional"/>
-    <xs:attribute name="Icon" type="xs:byte" use="required"/>
+    <xs:attribute name="Duration" type="xs:int" use="optional" />
+    <xs:attribute name="Tick" type="xs:int" use="optional"/>
+    <xs:attribute name="Icon" type="xs:unsignedShort" use="required"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Status.xsd]: I don't think icon can go beyond 255?

Comment (norrismiv) [reply HybrasylXML/XSD/Status.xsd]: Could be wrong, might go to 266.

File: HybrasylXML/XSD/Status.xsd [hunk 4]
```diff
@@ -38,14 +39,10 @@
...
   <xs:complexType name="CastRestrictions">
-    <xs:sequence minOccurs="0" maxOccurs="1">
-      <xs:element name="Use" type="status:CastRestriction"/>
-      <xs:element name="Receive" type="status:CastRestriction"/>
-    </xs:sequence>
+    <xs:attribute name="Use" type="xs:token"/>
+    <xs:attribute name="Heal" type="xs:token"/>
   </xs:complexType>
-
-  <xs:complexType name="CastRestriction"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Status.xsd]: If removing CastRestriction as a type, CastRestrictions should be singular.

File: HybrasylXML/XSD/Status.xsd [hunk 5]
```diff
@@ -76,14 +83,15 @@
           <xs:attribute name="Id" type="xs:unsignedByte" use="required" />
         </xs:complexType>
       </xs:element>
+      <xs:element name="Messages" minOccurs="0" maxOccurs="1" type="status:Messages"/>
```
Comment (norrismiv) [inline HybrasylXML/XSD/Status.xsd]: Message effects? ???????

File: examples/XML/creatures/Bee.xml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Creature xmlns="http://www.hybrasyl.com/XML/Creature" Sprite="1">
+  <Name>Bee</Name>
+  <Description>Ack!</Description>
```
Comment (norrismiv) [inline examples/XML/creatures/Bee.xml]: Pls fix description

File: examples/XML/creatures/Wolf.xml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Creature xmlns="http://www.hybrasyl.com/XML/Creature" Sprite="2">
+  <Name>Wolf</Name>
+  <Description>Has big plans to renovate Pravat</Description>
```
Comment (norrismiv) [inline examples/XML/creatures/Wolf.xml]: Just no

File: examples/XML/statuses/Blind.xml [hunk 8]
```diff
@@ -0,0 +1,22 @@
+<Status Duration="30" Tick="1" Icon="3" Name="Blind">
```
Comment (norrismiv) [inline examples/XML/statuses/Blind.xml]: These should not be under examples

File: examples/XML/statuses/Coma.xml [hunk 9]
```diff
@@ -0,0 +1,33 @@
+<Status Duration="30" Tick="1" Icon="24" Name="Coma">
```
Comment (norrismiv) [inline examples/XML/statuses/Coma.xml]: These should not be under examples

File: examples/XML/statuses/Freeze.xml [hunk 10]
```diff
@@ -0,0 +1,27 @@
+<Status Duration="30" Tick="1" Icon="50" Name="Freeze">
```
Comment (norrismiv) [inline examples/XML/statuses/Freeze.xml]: These should not be under examples.

File: examples/XML/statuses/Paralyze.xml [hunk 11]
```diff
@@ -0,0 +1,28 @@
+<Status Duration="30" Tick="1" Icon="97" Name="Paralyze">
```
Comment (norrismiv) [inline examples/XML/statuses/Paralyze.xml]: These should not be under examples.

File: examples/XML/statuses/Poison.xml [hunk 12]
```diff
@@ -0,0 +1,35 @@
+<Status Duration="30" Tick="1" Icon="97" Name="Poison">
```
Comment (norrismiv) [inline examples/XML/statuses/Poison.xml]: These should not be under examples

File: examples/XML/statuses/Sleep.xml [hunk 13]
```diff
@@ -0,0 +1,27 @@
+<Status Duration="30" Tick="1" Icon="2" Name="Sleep">
```
Comment (norrismiv) [inline examples/XML/statuses/Sleep.xml]: These should not be under examples

File: hybrasyl/CreatureStatus.cs [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+                Target.BonusRegen += effect.Regen;
+                Target.BonusMr += effect.Mr;
+                Target.BonusDamageModifier = effect.DamageModifier;
+                Target.BonusHealModifier = effect.HealModifier;
+                Target.BonusReflectChance += effect.ReflectChance;
+                Target.BonusReflectIntensity += effect.ReflectIntensity;
+                Target.BonusAc += effect.Str;
+                Target.OffensiveElementOverride = (Enums.Element)effect.OffensiveElement;
+                Target.DefensiveElementOverride = (Enums.Element)effect.OffensiveElement;
+
```
Comment (norrismiv) [inline hybrasyl/CreatureStatus.cs]: null handling? on effect properties?

File: hybrasyl/Inventory.cs [hunk 15]
```diff
@@ -84,8 +85,8 @@ public bool ConditionsValid
             {
                 return _source.Map == _target.Map && _source.IsInViewport(_target) &&
                        _target.IsInViewport(_source) &&
-                       _source.Status.HasFlag(PlayerCondition.InExchange) &&
-                       _target.Status.HasFlag(PlayerCondition.InExchange) &&
+                       _source.Condition.InExchange &&
+                       _target.Condition.InExchange &&
```
Comment (norrismiv) [inline hybrasyl/Inventory.cs]: should this not be an or statement between source and target?

File: hybrasyl/NumberCruncher.cs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    /// 
+
+    static class NumberCruncher
+    {
+
+        public static ILog Logger = LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
+        // This is dumb, but it's a consequence of how xsd2code works
+        static private double _evalSimple(dynamic simple)
+        {
```
Comment (norrismiv) [inline hybrasyl/NumberCruncher.cs]: private static for proper syntax.

File: hybrasyl/NumberCruncher.cs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+            }
+            catch (Exception e)
+            {
+                Logger.Error($"NumberCruncher formula error: castable {castable.Name}, target {target.Name}, source {source?.Name ?? "no source"}: {formula}, error: {e}");
+                return 0;
+            }
+        }
+
+        static public (Enums.DamageType Type, double Amount) CalculateDamage(Castable castable, Creature target, Creature source = null)
+        {
```
Comment (norrismiv) [inline hybrasyl/NumberCruncher.cs]: public static for proper syntax

File: hybrasyl/Objects/ConditionInfo.cs [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+                return conditionCheck;
+            }
+        }
+
+        public bool IsAttackable
+        {
+            get
+            {
+                if (User != null)
+                    return PvpEnabled;
```
Comment (norrismiv) [inline hybrasyl/Objects/ConditionInfo.cs]: This should also check for whether or not AscendEnabled.

AscendEnabled likely needs to be created.

File: hybrasyl/Objects/Creature.cs [hunk 19]
```diff
@@ -109,12 +221,272 @@ public Creature()
...
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(direction), direction, null);
+            }
+            if (obj is Creature) return obj as Creature;
+            return null;
+        }
+    
+        public virtual List<Creature> GetTargets(Castable castable, Creature target = null)
+        {
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: An array would likely be better in this instance

File: hybrasyl/Objects/Creature.cs [hunk 20]
```diff
@@ -330,42 +702,93 @@ public Creature LastHitter
...
+                        tar.Damage(damageOutput.Amount, OffensiveElement, damageOutput.Type, this);
+                    }
+                    else if (castObject.Effects?.Heal != null)
+                    {
+                        var healOutput = NumberCruncher.CalculateHeal(castObject, tar, this);
+                        tar.Heal(healOutput, this);
+                    }
+                    else if (castObject.Statuses?.Add != null)
+                    {
+
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: ??

File: hybrasyl/Objects/Creature.cs [hunk 21]
```diff
@@ -330,42 +702,93 @@ public Creature LastHitter
...
+                    }
+                    else if (castObject.Statuses?.Remove != null)
+                    {
+
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: ??

File: hybrasyl/Objects/Creature.cs [hunk 22]
```diff
@@ -557,23 +978,14 @@ public virtual void Motion(byte motion, short speed)
...
-            if (target.AbsoluteImmortal || target.PhysicalImmortal) return;
-            if (target.Hp == target.MaximumHp) return;
+            if (AbsoluteImmortal || PhysicalImmortal) return;
+            if (Hp == MaximumHp) return;
 
-            target.Hp = heal > uint.MaxValue ? target.MaximumHp : Math.Min(target.MaximumHp, (uint)(target.Hp + heal));
-            SendDamageUpdate(target);
-            if (target is User) { target.UpdateAttributes(StatUpdateFlags.Current); }
+            Hp = heal > uint.MaxValue ? MaximumHp : Math.Min(MaximumHp, (uint)(Hp + heal));
+            SendDamageUpdate(this);
```
Comment (norrismiv) [inline hybrasyl/Objects/Creature.cs]: this would work for self heal.. not healing a target

File: hybrasyl/Objects/Monster.cs [hunk 23]
```diff
@@ -627,19 +266,23 @@ public void AssailAttack(Direction direction, Creature target = null)
...
+            SimpleAttack(target);
+                            
             //animation handled here as to not repeatedly send assails.
             var assail = new ServerPacketStructures.PlayerAnimation() { Animation = 1, Speed = 20, UserId = this.Id };
-            var sound = new ServerPacketStructures.PlaySound() { Sound = (byte)1 };
             //Enqueue(assail.Packet());
             //Enqueue(sound.Packet());
             SendAnimation(assail.Packet());
-            PlaySound(sound.Packet());
+            PlaySound(1);
```
Comment (norrismiv) [inline hybrasyl/Objects/Monster.cs]: using the structure is proper for consistency, even if it is only a single byte returned.

File: hybrasyl/Objects/Monster.cs [hunk 24]
```diff
@@ -627,19 +266,23 @@ public void AssailAttack(Direction direction, Creature target = null)
...
         }
 
-
+        /// <summary>
+        /// A simple directional attack by a monster (equivalent of straight assail).
+        /// </summary>
+        /// <param name="direction"></param>
+        /// <param name="target"></param>
+        public void SimpleAttack(Creature target) => target?.Damage(_simpleDamage, OffensiveElement, Enums.DamageType.Physical, this);
```
Comment (norrismiv) [inline hybrasyl/Objects/Monster.cs]: was damage updated to account for element handling?

File: hybrasyl/Objects/User.cs [hunk 25]
```diff
@@ -611,10 +541,10 @@ public override void OnDeath()
...
         public void EndComa()
         {
-            if (!Status.HasFlag(PlayerCondition.InComa)) return;
-            ToggleNearDeath();
-            var bar = RemoveStatus(NearDeathStatus.Icon, false);
-            Logger.Debug($"EndComa: {Name}: removestatus for coma is {bar}");
+            if (!Condition.Comatose) return;
+            Condition.Comatose = false;
+            //var bar = RemoveStatus(NearDeathStatus.Icon, false);
+            Logger.Debug($"EndComa: {Name}: removestatus for coma is hurp");
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: durp

File: hybrasyl/Objects/User.cs [hunk 26]
```diff
@@ -1260,163 +1188,74 @@ public void SendVisibleGold(Gold gold)
...
-                }
-                if (spellCast) CastingCost(castable);
-
                 Client.Enqueue(new ServerPacketStructures.Cooldown()
                 {
                     Length = (uint)castable.Cooldown,
                     Pane = 1,
                     Slot = slot
-                }.Packet()); ;
+                }.Packet());
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: I really liked my extra ;

File: hybrasyl/Objects/User.cs [hunk 27]
```diff
@@ -1260,163 +1188,74 @@ public void SendVisibleGold(Gold gold)
...
             }
+            else
+                SendSystemMessage("It didn't work.");
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: "Failed." is the proper message

File: hybrasyl/Objects/User.cs [hunk 28]
```diff
@@ -1260,163 +1188,74 @@ public void SendVisibleGold(Gold gold)
...
-                }
+                    Length = (uint)castable.Cooldown,
+                    Pane = 1,
+                    Slot = slot
+                }.Packet());
             }
-
-            return true;
+            else
+                SendSystemMessage("It didn't work.");
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: "Failed." is the proper message.

File: hybrasyl/Objects/User.cs [hunk 29]
```diff
@@ -1260,163 +1188,74 @@ public void SendVisibleGold(Gold gold)
...
 
-            if (double.TryParse(costs.Stat.Hp, out double hpCostPercent))
-            {
-                if (hpCostPercent != 0)
-                {
-                    reduceHp = (uint)Math.Ceiling(this.MaximumHp * hpCostPercent);
-                }
-            }
+            // HP cost can be either a percentage (0.25) or a fixed amount (50)
+            if (costs.Stat.Hp.Contains('.'))
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: What if i want it to be 1, as in 1.00 without specifying 1.00 in xml?

Comment (norrismiv) [reply hybrasyl/Objects/User.cs]: the value taken when loaded will return as 1, even if specified in xml.

File: hybrasyl/Objects/User.cs [hunk 30]
```diff
@@ -2305,586 +2145,103 @@ public override void RegenerateMp(double mp, Creature regenerator = null)
...
-                            break;
-                    }
-
-                    if (!rect.IsEmpty) possibleTargets.AddRange(Map.EntityTree.GetObjects(rect).Where(obj => obj is Creature && obj != this && obj.GetType() != typeof(Merchant))); ;
-
-                    var actualTargets = intent.MaxTargets > 0 ? possibleTargets.Take(intent.MaxTargets).OfType<Creature>().ToList() : possibleTargets.OfType<Creature>().ToList();
-
-                    foreach (var target in actualTargets)
+                    Status applyStatus;
+                    if (World.WorldData.TryGetValue<Status>(status, out applyStatus))
```
Comment (norrismiv) [inline hybrasyl/Objects/User.cs]: applyStatus should be declared inline. https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7

File: hybrasyl/Objects/VisibleObject.cs [hunk 31]
```diff
@@ -215,11 +215,13 @@ public virtual void Effect(ushort effect, short speed)
             }
         }
 
-        public virtual void PlaySound(ServerPacket packet)
+        public virtual void PlaySound(byte Id)
         {
+            var soundPacket = new ServerPacketStructures.PlaySound() { Sound = Id };
+
```
Comment (norrismiv) [inline hybrasyl/Objects/VisibleObject.cs]: oh i see what you did

File: HybrasylXML/XSD/Castables.xsd (hunk 2 above)
Comment (baughj) [reply HybrasylXML/XSD/Castables.xsd]: `optional` here has no real effect in xsd2code, so I suppose we could drop it. Now that there is `None` it is less of a concern.

File: HybrasylXML/XSD/Castables.xsd (hunk 1 above)
Comment (baughj) [reply HybrasylXML/XSD/Castables.xsd]: To match extant datatypes (int/speed are not ints in Hybrasyl itself).

File: HybrasylXML/XSD/Status.xsd (hunk 3 above)
Comment (baughj) [reply HybrasylXML/XSD/Status.xsd]: thus, ushort.

File: HybrasylXML/XSD/Status.xsd (hunk 4 above)
Comment (baughj) [reply HybrasylXML/XSD/Status.xsd]: OK.

File: HybrasylXML/XSD/Status.xsd (hunk 5 above)
Comment (baughj) [reply HybrasylXML/XSD/Status.xsd]: A lot of basic statuses in usda have messages sent to the player. `The life is draining from your body` or `Poison is coursing through your veins`. Etc.

Comment (norrismiv) [reply HybrasylXML/XSD/Status.xsd]: I suppose that makes sense to have that tied in xml. OK

File: examples/XML/creatures/Bee.xml (hunk 6 above)
Comment (baughj) [reply examples/XML/creatures/Bee.xml]: done, lol

File: examples/XML/creatures/Wolf.xml (hunk 7 above)
Comment (baughj) [reply examples/XML/creatures/Wolf.xml]: done

File: hybrasyl/CreatureStatus.cs (hunk 14 above)
Comment (baughj) [reply hybrasyl/CreatureStatus.cs]: Everything here is a value type and cannot be null and/or has defaults assigned in XML.

File: hybrasyl/Inventory.cs (hunk 15 above)
Comment (baughj) [reply hybrasyl/Inventory.cs]: No - this is supposed to check that both participants of an exchange are still in it.

Comment (norrismiv) [reply hybrasyl/Inventory.cs]: OK

File: hybrasyl/CreatureStatus.cs (hunk 14 above)
Comment (norrismiv) [reply hybrasyl/CreatureStatus.cs]: OK

File: hybrasyl/NumberCruncher.cs (hunk 16 above)
Comment (baughj) [reply hybrasyl/NumberCruncher.cs]: Fixed

File: hybrasyl/NumberCruncher.cs (hunk 17 above)
Comment (baughj) [reply hybrasyl/NumberCruncher.cs]: Fixed

File: hybrasyl/Objects/ConditionInfo.cs (hunk 18 above)
Comment (baughj) [reply hybrasyl/Objects/ConditionInfo.cs]: Why does ascending need to be any different than having pvp enabled?

File: hybrasyl/NumberCruncher.cs (hunk 16 above)
Comment (norrismiv) [reply hybrasyl/NumberCruncher.cs]: note: this occurs in many places, just referenced a single for simplicity

File: hybrasyl/Objects/Creature.cs (hunk 19 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: Why?

File: hybrasyl/Objects/ConditionInfo.cs (hunk 18 above)
Comment (norrismiv) [reply hybrasyl/Objects/ConditionInfo.cs]: It doesn't need to be. I figure though it might be something better to have say if someone is PvpEnabled and also wants to Ascend. Having PvpEnabled should function differently than ascension in my opinion.

Example: PvpEnabled death should not allow the user to access Ascension related areas.

File: hybrasyl/Objects/Creature.cs (hunk 19 above)
Comment (norrismiv) [reply hybrasyl/Objects/Creature.cs]: In the case of a list desyncing due to creature spawn or death. Could be me overthinking it, but I can see the list being modified, and erroring. An array would not have this issue.

File: hybrasyl/Objects/Creature.cs (hunk 21 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: This is a consequence of having this defined in two places by necessity (monster versus user) and working mostly on the user - I will refactor it a bit and update.

Comment (norrismiv) [reply hybrasyl/Objects/Creature.cs]: 👍

File: hybrasyl/Objects/Creature.cs (hunk 22 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: `Heal` heals `this`. Healing another creature should be done via `othercreatureobj.Heal(3)`, not `this.Heal(3, othercreatureobj)`. Instances of healing elsewhere all use this pattern now and pass `source` as the healer object for future use.

File: hybrasyl/Objects/Monster.cs (hunk 23 above)
Comment (baughj) [reply hybrasyl/Objects/Monster.cs]: IMO the packet structure(s) should be used in the `Play*` or `Show*` functions themselves, not in individual functions (I changed the signature to take a numeric ID, as it did before). I don't believe that individual functions and handlers should need to construct packets themselves.

There is a lot of packet structure cleanup work that should be done, but I agree, using the structure, when it finally comes time to put something on the wire, is the correct pattern.

File: hybrasyl/Objects/Monster.cs (hunk 24 above)
Comment (baughj) [reply hybrasyl/Objects/Monster.cs]: No, not yet. I'll make a note to handle that.

File: hybrasyl/Objects/User.cs (hunk 25 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: This will be fixed

File: hybrasyl/Objects/User.cs (hunk 27 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: OK - `Failed.` handles a specific case in usda and this is more broad, but, fine.

File: hybrasyl/Objects/User.cs (hunk 28 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: see above

File: hybrasyl/Objects/User.cs (hunk 29 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: `1` would be the fixed value 1. `1.00` would be a percentage. It is a string in XML and a string type here. 

Since `1.36` has no meaning as a fixed value (given that HP/MP are `uint`) I'm not sure what the issue is here.

File: hybrasyl/Objects/User.cs (hunk 30 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: Huh. Neat! Will fix.

File: examples/XML/statuses/Blind.xml (hunk 8 above)
Comment (blueowlstudio) [reply examples/XML/statuses/Blind.xml]: Just a note, if the statuses are being kept, need to add xmlns="http://www.hybrasyl.com/XML/Status", otherwise server start generates "NullReferenceException" error.

File: examples/XML/maps/ChaoticThreshold.xml [hunk 32]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0"?>
+<Map xmlns="http://www.hybrasyl.com/XML/Maps" Id="300" X="30" Y="30">
+  <Name>Chaotic Threshold</Name>
+  <Npcs>
+    <Npc Name="Nylarlathotep" X="12" Y="6" Direction="0"/>
```
Comment (blueowlstudio) [inline examples/XML/maps/ChaoticThreshold.xml]: Nylarlathotep xml missing. Results in KeyNotFoundException on server start.

Comment (blueowlstudio) [inline examples/XML/maps/ChaoticThreshold.xml]: Nylarlathotep xml missing. Results in KeyNotFoundException on server start.

File: hybrasyl/Objects/User.cs (hunk 29 above)
Comment (blueowlstudio) [inline hybrasyl/Objects/User.cs]: May need a Null check here, as Assail does not currently have any stat costs, unless you want to make that a required element, with the Hp / Mp elements being optional.

File: hybrasyl/Objects/User.cs [hunk 33]
```diff
@@ -1260,163 +1188,74 @@ public void SendVisibleGold(Gold gold)
...
-                    {
-                        if (Inventory.Contains(item.Value, item.Quantity)) removeNumItems++;
-                    }
+                    if (Inventory.Contains(item.Value, item.Quantity)) removeNumItems++;
                 }
             }
 
+            // Check that all requirements are met first. Note that a spell cannot be cast if its HP cost would result
+            // in the caster's HP being reduced to zero.
+            if (reduceHp > Hp || reduceMp >= Mp || costs.Gold > Gold || removeNumItems != costs.Items.Count) return false;
```
Comment (blueowlstudio) [inline hybrasyl/Objects/User.cs]: Need null check for Items, unless you want to make this required element, with a default count of 0.

File: hybrasyl/Objects/Creature.cs [hunk 34]
```diff
@@ -109,12 +221,272 @@ public Creature()
...
+                            {
+                                //attack radius
+                                rect = new Rectangle(X - intent.Radius, Y - intent.Radius, Math.Max(intent.Radius, (byte)1), Math.Max(intent.Radius, (byte)1));
+                            }
+                            break;
+                    }
+
+                    if (!rect.IsEmpty)
+                    {
+                        var intersect = Map.EntityTree.GetObjects(Rect).Where(obj => obj is Creature && obj != this);
```
Comment (blueowlstudio) [inline hybrasyl/Objects/Creature.cs]: Rect should be the local variable rect?

File: hybrasyl/Objects/Creature.cs [hunk 35]
```diff
@@ -109,12 +221,272 @@ public Creature()
...
+                        // Disqualify targets based on intent
+                        // If we aren't targeting friendlies or pvp, remove all users entirely
+                        if (!intent.Target.Contains(IntentTarget.Friendly) && !intent.Target.Contains(IntentTarget.Pvp))
+                            intersect = intersect.Where(e => !(e is User));
+
+                        // If we aren't targeting hostiles, remove all monsters
+                        if (!intent.Target.Contains(IntentTarget.Hostile))
+                            intersect.Where(e => !(e is Monster || (e as Creature).Condition.PvpEnabled));
```
Comment (blueowlstudio) [inline hybrasyl/Objects/Creature.cs]: After this line, I don't see anything that adds the intersect targets to the possibleTargets List.

File: hybrasyl/Objects/User.cs [hunk 36]
```diff
@@ -2305,586 +2145,103 @@ public override void RegenerateMp(double mp, Creature regenerator = null)
...
-            {
-                //need to handle scripting
-            }
-        }
 
-        public List<Creature> PossibleTargets(Castable castable)
-        {
-            List<Creature> actualTargets = new List<Creature>();
+                if (castObject.Effects.Animations.OnCast.Target == null) continue;
+                Effect(castObject.Effects.Animations.OnCast.Target.Id, castObject.Effects.Animations.OnCast.Target.Speed);
```
Comment (moseleyc) [inline hybrasyl/Objects/User.cs]: Effect() only casts on the user that cast the spell, not on the target.

File: hybrasyl/Objects/User.cs [hunk 37]
```diff
@@ -2305,586 +2145,103 @@ public override void RegenerateMp(double mp, Creature regenerator = null)
...
                 }
-                catch (InvalidOperationException)
+                // Note that we ignore castables with both damage and healing effects present - one or the other.
+                // A future improvement might be to allow more complex effects.
+                else if (castObject.Effects.Heal != null)
                 {
-                    motion = castObject.Effects.Animations.OnCast.Player.FirstOrDefault(x => x.Class.Contains((Class)Class));
-
-                    Logger.ErrorFormat("{1}: contains more than one motion for a class definition, using first one found!", castObject.Name);
+                    var healOutput = NumberCruncher.CalculateHeal(castObject, tar, this);
```
Comment (moseleyc) [inline hybrasyl/Objects/User.cs]: Nothing appears to happen with healOutput after this value assignment.

File: examples/XML/statuses/Blind.xml (hunk 8 above)
Comment (baughj) [reply examples/XML/statuses/Blind.xml]: Yes, fixed.

File: examples/XML/maps/ChaoticThreshold.xml (hunk 32 above)
Comment (baughj) [reply examples/XML/maps/ChaoticThreshold.xml]: Fixed

File: hybrasyl/Objects/User.cs (hunk 36 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: Fixed, nice catch.

File: hybrasyl/Objects/User.cs (hunk 37 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: Fixed, pebkac

File: hybrasyl/Objects/Creature.cs (hunk 34 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: Fixed, another nice catch

File: hybrasyl/Objects/Creature.cs (hunk 35 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: This has been refactored (and verified to work).

File: hybrasyl/Objects/User.cs (hunk 33 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: Fixed.

File: hybrasyl/Objects/User.cs (hunk 29 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: Fixed, also cleaning up a lot of nullrefs elsewhere

**Comment by norrismiv** [comment]
👍

File: HybrasylXML/XSD/hybrasylExtensions.cs [hunk 38]
```diff
@@ -173,16 +184,63 @@ public int Id
...
         public byte CastableLevel { get; set; }
 
+        public DateTime LastCast { get; set; }
+
+        public bool OnCooldown
+        {
+            get
+            {
+                if (Cooldown > 0)
+                    return (DateTime.Now - LastCast).Seconds > Cooldown;
```
Comment (moseleyc) [inline HybrasylXML/XSD/hybrasylExtensions.cs]: Should this be a less-than sign instead? If number of seconds passed is LT the cooldown, then false....

File: hybrasyl/Objects/User.cs [hunk 39]
```diff
@@ -1260,163 +1136,95 @@ public void SendVisibleGold(Gold gold)
...
+                SpellBook[slot].LastCast = DateTime.Now;
+                Client.Enqueue(new ServerPacketStructures.Cooldown()
                 {
-                    if (item != null)
-                    {
-                        if (!(Inventory.Contains(item.Value, item.Quantity))) return false;
-                    }
-                }
+                    Length = (uint)castable.Cooldown,
+                    Pane = 1,
```
Comment (moseleyc) [inline hybrasyl/Objects/User.cs]: I believe Pane = 0 for Spell's.

File: hybrasyl/Objects/Creature.cs [hunk 40]
```diff
@@ -330,42 +722,215 @@ public Creature LastHitter
...
+
+                // Handle statuses
+
+                foreach (var status in castObject.Effects.Statuses.Add)
+                {
+                    Status applyStatus;
+                    if (World.WorldData.TryGetValueByIndex<Status>(status.Value, out applyStatus))
+                    {
+                        ActivityLogger.Info($"UseCastable: {Name} casting {castObject.Name} - applying status {status.Value}");
+                        ApplyStatus(new CreatureStatus(applyStatus, tar, castObject));
```
Comment (moseleyc) [inline hybrasyl/Objects/Creature.cs]: Does duration / tick get passed through a different means? When I tried it, I only got the default.

File: hybrasyl/Objects/User.cs (hunk 39 above)
Comment (cmw-mnorris) [reply hybrasyl/Objects/User.cs]: Correct, spell pane is 0

File: HybrasylXML/XSD/hybrasylExtensions.cs (hunk 38 above)
Comment (baughj) [reply HybrasylXML/XSD/hybrasylExtensions.cs]: Fixed

File: hybrasyl/Objects/Creature.cs (hunk 40 above)
Comment (baughj) [reply hybrasyl/Objects/Creature.cs]: Override works now.

File: hybrasyl/Objects/User.cs (hunk 39 above)
Comment (baughj) [reply hybrasyl/Objects/User.cs]: Fixed.
